Sources/SwiftyUserDefaults.swift

```diff
@@ -467,33 +442,66 @@ extension UserDefaults {
 // MARK: NSCoding
 
 extension UserDefaults {
-    // TODO: Can we simplify this and ensure that T is NSCoding compliant?
     
-    public func archive<T>(_ key: DefaultsKey<T>, _ value: T) {
-        if let value: AnyObject = value as? AnyObject {
-            set(key, NSKeyedArchiver.archivedData(withRootObject: value))
-        } else {
-            assertionFailure("Invalid value type, needs to be a NSCoding-compliant type")
+    public func archive<T: NSCoding>(_ key: DefaultsKey<T>, _ value: T) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<T?>, _ value: T?) {
+        guard let value = value else {
+            remove(key)
+            return
         }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[T]>, _ value: [T]) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
     }
     
-    public func archive<T>(_ key: DefaultsKey<T?>, _ value: T?) {
-        if let value: AnyObject = value as? AnyObject {
-            set(key, NSKeyedArchiver.archivedData(withRootObject: value))
-        } else if value == nil {
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[T]?>, _ value: [T]?) {
+        guard let value = value else {
             remove(key)
-        } else {
-            assertionFailure("Invalid value type, needs to be a NSCoding-compliant type")
+            return
         }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[String: T]>, _ value: [String: T]) {
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
     }
     
-    public func unarchive<T>(_ key: DefaultsKey<T?>) -> T? {
+    public func archive<T: NSCoding>(_ key: DefaultsKey<[String: T]?>, _ value: [String: T]?) {
+        guard let value = value else {
+            remove(key)
+            return
+        }
+        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
+    }
+    
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<T?>) -> T? {
         return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? T
     }
     
-    public func unarchive<T>(_ key: DefaultsKey<T>) -> T? {
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<T>) -> T? {
         return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? T
     }
+    
+    public func unarchive<T: NSCoding>(_ key: DefaultsKey<[T]?>) -> [T]? {
+        return data(forKey: key._key).flatMap { NSKeyedUnarchiver.unarchiveObject(with: $0) } as? [T]
```

Comment:
The reason it worked is because I didn't make conformance to NSCoding a requirement on T — Swift Array and Dictionary structs aren't. And so in this case, the array would just automatically be bridged to NSArray which will archive fine.

I know it's pretty terrible, and your solution is probably better long term, but I don't want to stack these changes together...

This code, slightly changed for Swift 3 seems to work fine:

~~~swift
    // TODO: Can we simplify this and ensure that T is NSCoding compliant?
    
    public func archive<T>(_ key: DefaultsKey<T>, _ value: T) {
        set(key, NSKeyedArchiver.archivedData(withRootObject: value))
    }
    
    public func archive<T>(_ key: DefaultsKey<T?>, _ value: T?) {
        if let value = value {
            set(key, NSKeyedArchiver.archivedData(withRootObject: value))
        } else {
            remove(key)
        }
    }
~~~